nemo/collections/llm/perceiver/model/perceiver_layer.py

```diff
@@ -0,0 +1,229 @@
+        )
+
+    def forward(
+        self,
+        latent_states: torch.Tensor,
+        input_sequence: torch.Tensor,
+        cross_attention_masks: torch.Tensor = None,
+        self_attention_mask: torch.Tensor = None,
+        inference_params = None,
+    ):
+        """Forward pass through a Perceiver layer.
+
+        Args:
+            latent_states (torch.Tensor): Latent array of shape [latent_len, batch_size, hidden_size]
+                that learns to attend to the input sequence.
+            input_sequence (torch.Tensor): Input sequence of shape [seq_len, batch_size, hidden_size]
+                containing the actual input data.
+            cross_attention_masks (torch.Tensor): Tuple of attention masks (query_mask, key_value_mask) for
+                cross-attention. Each mask has shape [batch_size, 1, 1, seq_len].
+            self_attention_mask (torch.Tensor): Attention mask for self-attention on latents,
+                with shape [batch_size, 1, 1, latent_len].
+            inference_params (Optional[dict]): Optional parameters to optimize inference-time computation.
+
+        Returns:
+            torch.Tensor: Updated latent states after processing through the layer,
+                with shape [latent_len, batch_size, hidden_size].
+        """
+        print("\nShape tracking through forward pass:")
+        print(f"Initial latent_states: {latent_states.shape}")  # [latent_len, batch_size, hidden_size]
+        print(f"Input sequence: {input_sequence.shape}")  # [seq_len, batch_size, hidden_size]
+
+        # Cross attention from latents to input sequence
+        cross_attn_output = self.input_cross_attention(
+            latent_states,
+            key_value_states=input_sequence,
+            attention_mask=cross_attention_masks,
+            inference_params=inference_params,
+        )
+        print(f"After cross attention - output: {cross_attn_output[0].shape}, bias: {cross_attn_output[1].shape}")
+        latent_states = cross_attn_output
+
+        # Multiple self-attention layers
+        for i, self_attn_layer in enumerate(self.self_attention_layers):
+            self_attn_output = self_attn_layer(
+                latent_states[0],
+                attention_mask=self_attention_mask,
+                inference_params=inference_params,
+            )
+            print(f"After self attention layer {i} - output: {self_attn_output[0].shape}, bias: {self_attn_output[1].shape}")
+            latent_states = self_attn_output
+
+        latent_states = self.final_layernorm(latent_states[0])
+        print(f"After final layernorm: {latent_states.shape}")
+
+        output = make_viewless_tensor(
+            inp=latent_states,
+            requires_grad=latent_states.requires_grad,
+            keep_graph=True
+        )
+        print(f"Final output: {output.shape}")
+
+        return output
+
+
+# Example ModuleSpec for Perceiver layer
+perceiver_layer_spec = ModuleSpec(
```

Comment:
Do you think it would be useful to have a non-TE option as well? If so, it would be good to guard the TE imports.

Reply:
Any reason? We haven't been diligent to import-guard TE in the rest of the framework; our other models (ESM-2, Geneformer, etc.) have it as a hard dependency)